File: 剑指offer/47.礼物的最大价值.kt
class Solution {
  fun maxValue(grid: Array<IntArray>): Int {
    // 指定dp[i][j]为到达(i,j)位置的最大礼物值，他只能从上方下来或从左方过来
    val dp = Array(grid.size){IntArray(grid[0].size)}
    for(i in grid.indices){
      for(j in grid[0].indices){
        if(i==0 && j==0)
          dp[0][0]=grid[0][0]
        else
          if(i==0) dp[i][j]=dp[0][j-1]+grid[i][j]
          else if(j==0) dp[i][j]=dp[i-1][j]+grid[i][j]
          else dp[i][j]=maxOf(dp[i-1][j],dp[i][j-1])+grid[i][j]
      }
    }
    return dp.last().last();
  }
}

File: 剑指offer/47. 礼物的最大价值.py
class Solution:
    def maxValue(self, grid: List[List[int]]) -> int:
        # 奇数行，可以全拿
        # 偶数行，奇数列也能全拿
        m,n = len(grid),len(grid[0])
        dp = [[0]*n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                dp[i][j]=grid[i][j]
                if i and j:
                    dp[i][j]=max(dp[i-1][j],dp[i][j-1])+grid[i][j]
                elif i:
                    dp[i][j]=dp[i-1][j]+grid[i][j]
                elif j:
                    dp[i][j]=dp[i][j-1]+grid[i][j]
                else:
                    dp[i][j]=grid[i][j]
        return dp[-1][-1]